commands/bootstrap

```diff
@@ -297,6 +332,331 @@ create_self_signed() {
+
+    if ! "${EXEC_CMD[@]}" "$TARGET" test -f "$LOCAL_CERT_KEY"; then
+        info "Creating device private key"
+        "${EXEC_CMD[@]}" "$TARGET" $SUDO openssl genrsa -out "$LOCAL_CERT_KEY" 2048
+        "${EXEC_CMD[@]}" "$TARGET" sudo chown tedge:root "$LOCAL_CERT_KEY"
+        "${EXEC_CMD[@]}" "$TARGET" sudo chmod 400 "$LOCAL_CERT_KEY"
+    else
+        info "Using existing private key"
+    fi
+
+    # Create Certificate Signing Request (CSR)
+    DEVICE_CSR=$(
+        "${EXEC_CMD[@]}" "$TARGET" $SUDO openssl req \
+            -key "$LOCAL_CERT_KEY" \
+            -new \
+            -subj "/O=thin-edge/OU=${DEVICE_ID}/CN=${SERVICE}"
+    )
+
+    # Sign the CSR
+    if [ "$SERVICE" = "child-agent" ]; then
+    CERT_EXT=$(cat << EOF
+basicConstraints=CA:FALSE
+extendedKeyUsage = clientAuth
+EOF
+    )
+    else
+    CERT_EXT=$(cat << EOF
+authorityKeyIdentifier=keyid
+basicConstraints=CA:FALSE
+keyUsage = digitalSignature, keyAgreement
+extendedKeyUsage = serverAuth, clientAuth
+subjectAltName = DNS:${TARGET_DNS},IP:${TARGET_IP},DNS:localhost,IP:127.0.0.1
+EOF
+    )
+    fi
+
+    DEVICE_CERT_CHAIN="/tmp/${LOCAL_CERT_NAME}.crt"
+    openssl x509 -req \
+        -in <(echo "$DEVICE_CSR") \
+        -out "$DEVICE_CERT_CHAIN" \
+        -CA "$SIGNING_CERT_FILE" \
+        -CAkey "$SIGNING_CERT_KEY" \
+        -extfile <(echo "$CERT_EXT") \
+        -CAcreateserial \
+        -days "$LEAF_EXPIRE_DAYS"
+
+    # Build certificate chain (from leaf cert to the signing cert)
+    cat "$SIGNING_CERT_FILE" >> "$DEVICE_CERT_CHAIN"
+
+    # Copy cert to device
+    "${SCP_CMD}" "$DEVICE_CERT_CHAIN" "$TARGET" "/tmp/${LOCAL_CERT_NAME}.crt"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" mv "/tmp/${LOCAL_CERT_NAME}.crt" "$LOCAL_CERT_FILE"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" chown tedge:root "$LOCAL_CERT_FILE"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" chmod 644 "$LOCAL_CERT_FILE"
+
+    case "$SERVICE" in
+        mqtt-broker)
+            configure_mtls_mqtt_broker "$TARGET_IP" "$TARGET_DNS" "$LOCAL_SIGNING_CERT_FILE" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        main-agent)
+            configure_mtls_mqtt_client "$TARGET_IP" "$TARGET_DNS" "$LOCAL_SIGNING_CERT_FILE" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            configure_mtls_main_agent "$TARGET_IP" "$TARGET_DNS" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        c8y-mapper)
+            configure_mtls_c8y_mapper "$TARGET_IP" "$TARGET_DNS" "$LOCAL_CERT_FILE" "$LOCAL_CERT_KEY"
+            ;;
+        child-agent)
+	    # GET IP and DNS of the main device
+	    MAIN_DEVICE_DNS=$("${EXEC_CMD[@]}" "$MAIN_DEVICE" hostname 2>/dev/null || true)
+            if [ -z "$MAIN_DEVICE_DNS" ]; then
+                MAIN_DEVICE_DNS=$(echo "$MAIN_DEVICE" | sed -E 's/.*@(\S+)/\1/')
+            fi
+            MAIN_DEVICE_IP=$("${EXEC_CMD[@]}" "$MAIN_DEVICE" ip route get 8.8.8.8 2>/dev/null | sed -E 's/.*src (\S+) .*/\1/;t;d' 2>/dev/null || true)
```

Comment:
I'm not 100% sure using the public dns server 8.8.8.8 will work if the device has two network adapters, I suspect this would only return the network adapter that can reach the internet, where as a local device will most likely be connected via the internal (control) network, so we might need to check from the child device's perspective the ip address of the main device to ensure we get an ip address which is reachable from the child device.

Reply:
Technically we could avoid this by using the `0.0.0.0` bind address, however this would be bad security practice (and I guess is the reason why you're looking for an explicit ip address in the first place)...

Though mosquitto does support binding to an interface rather than an explicit adapter (via `bind_interface`), however I'm not sure if thin-edge.io supports this.

Reply:
> I'm not 100% sure using the public dns server 8.8.8.8 will work if the device has two network adapters, I suspect this would only return the network adapter that can reach the internet, where as a local device will most likely be connected via the internal (control) network, so we might need to check from the child device's perspective the ip address of the main device to ensure we get an ip address which is reachable from the child device.

Correct.

> Technically we could avoid this by using the `0.0.0.0` bind address, however this would be bad security practice (and I guess is the reason why you're looking for an explicit ip address in the first place)...

Indeed, a bad security practice.

> Though mosquitto does support binding to an interface rather than an explicit adapter (via `bind_interface`), however I'm not sure if thin-edge.io supports this.

`tedge config` provides a setting for `mqtt.external.bind.interface`.